Mathlib/RingTheory/TwoSidedIdeal.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.RingTheory.Congruence
+import Mathlib.RingTheory.Ideal.Basic
+import Mathlib.Algebra.Module.LinearMap.Basic
+import Mathlib.Algebra.Module.Opposites
+
+/-!
+# Two Sided Ideals
+
+In this file, for any `Ring R`, we reinterpret `I : RingCon R` as a two sided ideal of a ring.
+
+## Notes
+`SetLike (RingCon R) R` makes sense for any `NonUnitalNonAssocRing R`.
+But later for the `module` part, we will assume `R` is a ring.
+
+-/
+
+section ideal
+
+namespace RingCon
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] (I : RingCon R)
+
+instance : SetLike (RingCon R) R where
+  coe t := {r | t r 0}
+  coe_injective' t₁ t₂ (h : {x | _} = {x | _}) := by
+    refine RingCon.ext fun a b ↦ ⟨fun H ↦ ?_, fun H ↦ ?_⟩
+    · have H' : a - b ∈ {x | t₁ x 0} := sub_self b ▸  t₁.sub H (t₁.refl b)
+      rw [h] at H'
+      convert t₂.add H' (t₂.refl b) using 1 <;> abel
+    · have H' : a - b ∈ {x | t₂ x 0} := sub_self b ▸  t₂.sub H (t₂.refl b)
+      rw [← h] at H'
+      convert t₁.add H' (t₁.refl b) using 1 <;> abel
+
+lemma mem_iff (x : R) : x ∈ I ↔ I x 0 := Iff.rfl
+
+lemma rel_iff (x y : R) : I x y ↔ (x - y) ∈ I := by
+  rw [mem_iff]
+  constructor
+  · intro h; convert I.sub h (I.refl y); abel
+  · intro h; convert I.add h (I.refl y) <;> abel
+
+lemma le_iff (I J : RingCon R) : I ≤ J ↔ (I : Set R) ⊆ (J : Set R) := by
+  constructor
+  · intro h x hx
+    exact h hx
+  · intro h x y hxy
+    have h' : J _ _ := h <| sub_self y ▸ I.sub hxy (I.refl y)
+    convert J.add h' (J.refl y) <;>
+    abel
+
+lemma lt_iff (I J : RingCon R) : I < J ↔ (I : Set R) ⊂ (J : Set R) := by
+  rw [lt_iff_le_and_ne, Set.ssubset_iff_subset_ne, le_iff]
+  simp
+
+lemma zero_mem : 0 ∈ I := I.refl 0
+
+lemma add_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x + y ∈ I := by
+  simpa using I.add hx hy
+
+lemma neg_mem {x} (hx : x ∈ I) : -x ∈ I := by
+  simpa using I.neg hx
+
+lemma sub_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x - y ∈ I := by
+  simpa using I.sub hx hy
+
+lemma mul_mem_left (x y) (hy : y ∈ I) : x * y ∈ I := by
+  simpa using I.mul (I.refl x) hy
+
+lemma mul_mem_right (x y) (hx : x ∈ I) : x * y ∈ I := by
+  simpa using I.mul hx (I.refl y)
+
+lemma nsmul_mem {x} (n : ℕ) (hx : x ∈ I) : n • x ∈ I := by
+  simpa using I.nsmul _ hx
+
+lemma zsmul_mem {x} (n : ℤ) (hx : x ∈ I) : n • x ∈ I := by
+  simpa using I.zsmul _ hx
+
+instance : AddSubgroupClass (RingCon R) R where
+  zero_mem := zero_mem
+  add_mem := @add_mem _ _
+  neg_mem := @neg_mem _ _
```

Comment:
Can you add two [SMulMemClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/SubMulAction.html#SMulMemClass) instances with R and R^mop respectively?

Reply:
done in here [ee508d5](https://github.com/leanprover-community/mathlib4/pull/13401/commits/ee508d56d6c46d16bf283fc368962d9ba7dddc34)